pkg/liquidity-mining/test/L2GaugeCheckpointer.test.ts

```diff
@@ -0,0 +1,223 @@
+      })
+    );
+
+    // Gauge adder & add factories to gauge adder.
+    gaugeAdder = await deploy('GaugeAdder', { args: [gaugeController.address, ZERO_ADDRESS] });
+    const action = await actionId(gaugeAdder, 'addGaugeFactory');
+    await vault.grantPermissionsGlobally([action], admin);
+
+    await Promise.all(
+      gaugeFactories.map((factory) => gaugeAdder.connect(admin).addGaugeFactory(factory.contract.address, factory.type))
+    );
+
+    // Create some gauges from each factory.
+    await Promise.all(
+      gaugeFactories.map(async (factory) =>
+        gauges.set(factory.type, await createGauges(factory.contract, factory.type, GAUGES_PER_TYPE))
+      )
+    );
+  });
+
+  before('get test gauges', () => {
+    testGaugeType = GaugeType.Polygon;
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    testGauges = gauges.get(testGaugeType)!;
+  });
+
+  sharedBeforeEach('deploy L2 gauge checkpointer', async () => {
+    L2GaugeCheckpointer = await deploy('L2GaugeCheckpointer', {
+      args: [gaugeAdder.address],
+    });
+  });
+
+  describe('add gauges', () => {
+    context('with invalid gauge type', () => {
+      it('reverts', async () => {
+        for (const gaugeType of UNSUPPORTED_GAUGE_TYPES) {
+          await expect(L2GaugeCheckpointer.addGauges(gaugeType, testGauges)).to.be.revertedWith(
+            'Unsupported gauge type'
+          );
+        }
+      });
+    });
+
+    context('with incorrect factory and controller setup', () => {
+      it('reverts', async () => {
+        const otherGaugeType = GaugeType.Optimism;
+        await expect(L2GaugeCheckpointer.addGauges(otherGaugeType, testGauges)).to.be.revertedWith(
+          'Gauge does not come from valid factory'
+        );
+      });
+    });
+
+    context('with correct factory and wrong controller setup', () => {
+      it('reverts', async () => {
+        await expect(L2GaugeCheckpointer.addGauges(testGaugeType, testGauges)).to.be.revertedWith(
+          'Gauge does not exist in controller'
+        );
+      });
+    });
+
+    context('with correct factory and controller setup', () => {
+      sharedBeforeEach('add gauges to regular controller', async () => {
+        await addGaugesToController(gaugeController, testGauges);
+      });
+
+      it('adds stakeless gauges correctly using only the correct gauge type', async () => {
+        const tx = await L2GaugeCheckpointer.addGauges(testGaugeType, testGauges);
+        const receipt = await tx.wait();
+        for (const testGauge of testGauges) {
+          expectEvent.inReceipt(receipt, 'GaugeAdded', { gaugeType: testGaugeType, gauge: testGauge });
+        }
+        await expectGaugesAdded(testGaugeType, testGauges);
+        // Check that all remaining types are empty.
+        await Promise.all(
+          GAUGE_TYPES.filter((gaugeType) => gaugeType != testGaugeType).map((gaugeType) =>
+            expectGaugesAdded(gaugeType, [])
+          )
+        );
+      });
+
```

Comment:
You can split this into three tests: one where you check the gauges have been added (i.e. the contract returns true in `hasGauge` for all of them), one where you test the events (`it('emits one event per added gauge')`), and one where you enumerate the gauges and find all of them. You'd start all three `it`s with `await checkpointer.adddGauges(...)`.

Reply:
+1; done.